docs/design/unsupported-args-design.md

```diff
@@ -0,0 +1,118 @@
+# Unsupported Arguments schema and common design
+Date: 2024-06-29
+
+## Abstract
+OADP allows to configure it's components parameters via number of DataProtectionApplication Schema.
+
+Unsupported arguments within the OADP operator are those that do not conform to the standard DataProtectionApplication Schema or are explicitly identified as unsupported or custom arguments.
+
+There is need to provide a common way for defining and managing unsupported arguments within the OADP operator for its' components such as velero server or node-agent arguments.
+
+## Background
+The OADP allows configuration of several predefined arguments via the DataProtectionApplication Schema. These parameters are carefully selected and tested to meet the highest quality standards. However, in some circumstances, there is a need to pass other parameters that may be under development, not yet fully tested, or specifically overriding the default ones.
+
+Over the past OADP releases, some implementations have provided access to extra arguments. However, there has been no standardization, and unknown to DataProtectionApplication Schema arguments were not allowed.
+
+## Goals
+- Define common way to allow passing extra arguments to various executables used by OADP.
+- Ensure the new common way do not interfere with already existing implementations.
+
+## Non Goals
+- Deprecate current ways of defining extra arguments.
+- Change precedences within current ways of defining extra arguments.
+- Validation of unsupported args, they are passed to the appropriate executables without any prior validation.
+
+## Design
+Unsupported arguments are defined as a `ConfigMap` within the same Namespace as OADP Operator.
+
+There may be multiple `ConfigMaps` for particular section of the OADP configuration.
+
+There are default naming convention for the `ConfigMaps`, however user may override those using `DPA` annotations.
+
+When ConfigMap is found, controller takes those values as the highest priority and replaces **all** of the other configuration options, even if they were defined within `DPA` schema for the section for which the `ConfigMap` was created.
+
+The Schema of an `ConfigMap` includes data that corresponds to the arguments passed to the relevant executable as in the below example:
+
+   ```
+   apiVersion: v1
+   kind: ConfigMap
+   metadata:
+     name: oadp-unsupported-velero-server-args
+     namespace: openshift-adp
+   data:
+     --default-volume-snapshot-locations: 'aws:backups-primary,azure:backups-secondary'
```

Comment:
Are there any argument names that won't be acceptable as ConfigMap keys? I tried some different strings with `/` and `:` characters, and got an error message that says `regex used for validation is '[-._a-zA-Z0-9]+'`.

I'm not sure how restrictive this will be in practice, but it might be worth considering in case any of our binaries have any arguments that won't work this way.

Reply:
I think we can poke round how cobra works (velero CLI is built using it), if it also has the same restrictions

Reply:
tried and I can have a flag name like `name/flag` :grimacing: BUT today, I think there is none like this (would say velero flags name are composed of  lowercase letters, `-` and `_`

Maybe we can enforce this regex in velero flags names using some cobra functionalities

Reply:
The K8s doesn't restrict special characters in the `value` part, however the key must be within regex as @mrnold wrote:

```
Invalid value: "--default-volume-snapshot_/locations": a valid config key must consist of alphanumeric characters, '-', '_' or '.' (e.g. 'key.name',  or 'KEY_NAME',  or 'key-name', regex used for validation is '[-._a-zA-Z0-9]+')
``` 

If we ever want to have such special args, which I don't think it's a valid use case at the moment - the args are within the regex imo, we can not translate them into base64 encodings as that may produce `=` chars, what we could do is to have list of args numerated in configmap and have configmap annotated that it's using this method?, e.g.:

Instead of:
```yaml
   apiVersion: v1
   kind: ConfigMap
   metadata:
     name: oadp-unsupported-velero-server-args
     namespace: openshift-adp
    annotations:
      numerated-args: "true"
data:
   --my-super&strange/arg: 'arg value arg second value'
   --my-second-arg: 'arg value 2'
```
```yaml
  apiVersion: v1
  kind: ConfigMap
  metadata:
    name: oadp-unsupported-velero-server-args
    namespace: openshift-adp
    annotations:
      numerated-args: "true"
data:
   arg-1: '--my-super&strange/arg: arg value arg second value'
   arg-2: '--my-second-arg: 'arg value 2'
```

Reply:
I am closing this as it's not relevant anymore.